### PR TITLE
Make it possible to load a pre-compiled state

### PR DESCRIPTION
**Pull request by kopyl**
I made it possible to load a pre-compiled state to avoid long cold-starts.

Addressing my issue: https://github.com/Oneflow-Inc/diffusers/issues/362.

Please merge.

### CODE REVIEW

File: src/onediff/infer_compiler/with_oneflow_compile.py [hunk 1]
```diff
@@ -304,3 +304,27 @@ def wrap_module(module):
     model = wrap_module(torch_module)
     model._register_state_dict_hook(state_dict_hook)
     return model
+
+
+def oneflow_load_compiled(torch_module, path, *, device=None, use_graph=True, options={}):
```
Comment (strint) [inline src/onediff/infer_compiler/with_oneflow_compile.py:309]: oneflow_load_compiled seems lilke a nice syntactic sugar for your case.

But adding this  API to onediff is not that nice for the moment. Small and composable APIs are preferred, and compiling joint with loading is a little complicated.

I think you could add the function to your project's utils lib to test its usability.

Comment (kopyl) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: > Small and composable APIs are preferred
Like what? How do you currently achieve the same functionality? How others are supposed to know how to load a compiled unet?

> compiling joint with loading
Could you please explain what does it mean here?

Also i'm wonderting if it's possible to load compiled unet directly to a Diffusers pipeline without loading an original unet first...

Comment (strint) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: > > Small and composable APIs are preferred
> 
> Like what? How do you currently achieve the same functionality? How others are supposed to know how to load a compiled unet?

Here is a wiki and example on compiling and loading. 
- Save/load/change device of the compiled graph in https://github.com/Oneflow-Inc/diffusers/wiki/OneDiff-v0.11-and-supporting-for-Stable-Diffusion-XL

> > compiling joint with loading
> 
> Could you please explain what does it mean here?

Usually we need to compile a torch module to a graph with `oneflow_compile`; but after `oneflow_compile` the graph hasn't done any optimization yet.

At your first call on the graph( compiled unet), the graph will do optimization depending on the inputs of the graph. This is why the first call takes a little longer. After the first call, no compiling is need until the graph got new inputs(input shape is changed).

The first optimization takes about 1 minute (on my machine). If you want to save the optimization time, loading an optimizaion result with `warmup_with_load` is a good choice. Loading only takes about 2 seconds.

> Also i'm wonderting if it's possible to load compiled unet directly to a Diffusers pipeline without loading an original unet first...

The Diffusers Pipeline is loading model state_dict(the model weights), so the unet in pipeline is also loading state_dict.

But The compiled unet is loading the optimization result, and the optimization result has no model weights in it. 

So there is no waste when loading the optimization result after the pipeline has loaded the model weights.

Comment (kopyl) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: @strint thanks. By the way, why `load_graph` and `warmup_with_load` are the same?

```
def load_graph(self, file_path, device=None):
    self.get_graph().warmup_with_load(file_path, device)

def warmup_with_load(self, file_path, device=None):
    self.get_graph().warmup_with_load(file_path, device)
```

Comment (strint) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: > why load_graph and warmup_with_load are the same

warmup_with_load is the old API, it stays here just for forward compatibility.

Comment (strint) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: I think you are right, we should delete `warmup_with_load`.

Comment (kopyl) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: @strint thanks. I did not know there is such a thing as forward compatibility. There is usually a backward one.

Comment (strint) [reply src/onediff/infer_compiler/with_oneflow_compile.py:309]: > @strint thanks. I did not know there is such a thing as forward compatibility. There is usually a backward one.

You are right, warmup_with_load is for backward compatibility.  There is forward compatibility, but it dosen't matter here.

**Pull request closed without merging**